src/client/error.rs

```diff
@@ -0,0 +1,264 @@
+    BulkWriteError(BulkWriteException),
+    ArgumentError(String),
+    OperationError(String),
+    ResponseError(String),
+    CursorMissingError,
+    LockError,
+    DefaultError(String),
+}
+
+/// The error type for Write-related MongoDB operations.
+#[derive(Debug)]
+pub struct WriteException {
+    kind: WriteExceptionKind,
+    code: i32,
+    message: String,
+}
+
+/// The two types of single-write related MongoDB operations.
+#[derive(Debug)]
+pub enum WriteExceptionKind {
+    WriteConcernError(bson::Document),
+    WriteError,
+}
+
+/// The error struct for Bulk-Write related MongoDB operations.
+#[derive(Debug)]
+pub struct BulkWriteException {
+    processed_requests: Vec<WriteModel>,
+    unprocessed_requests: Vec<WriteModel>,
+    write_concern_error: Option<WriteException>,
+    write_errors: Vec<BulkWriteError>,
+    message: String,
+}
+
+/// The error struct for a single bulk-write step, indicating the request
+/// and its index in the original bulk-write request.
+#[derive(Debug)]
+pub struct BulkWriteError {
+    index: i32,
+    request: Option<WriteModel>,
+}
+
+impl<'a> From<&'a str> for Error {
+    fn from(s: &str) -> Error {
+        Error::DefaultError(s.to_owned())
+    }
+}
+
+impl From<String> for Error {
+    fn from(s: String) -> Error {
+        Error::DefaultError(s.to_owned())
+    }
+}
+
+impl From<WriteException> for Error {
+    fn from(err: WriteException) -> Error {
+        Error::WriteError(err)
+    }
+}
+
+impl From<BulkWriteException> for Error {
+    fn from(err: BulkWriteException) -> Error {
+        Error::BulkWriteError(err)
+    }
+}
+
+impl From<bson::EncoderError> for Error {
+    fn from(err: bson::EncoderError) -> Error {
+        Error::EncoderError(err)
+    }
+}
+
+impl From<bson::DecoderError> for Error {
+    fn from(err: bson::DecoderError) -> Error {
+        Error::DecoderError(err)
+    }
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Error {
+        Error::IoError(err)
+    }
+}
+
+impl From<byteorder::Error> for Error {
+    fn from(err: byteorder::Error) -> Error {
+        Error::IoError(From::from(err))
+    }
+}
+
+impl<T> From<sync::PoisonError<T>> for Error {
+    fn from(_: sync::PoisonError<T>) -> Error {
+        Error::LockError
+    }
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            &Error::WriteError(ref inner) => inner.fmt(fmt),
+            &Error::BulkWriteError(ref inner) => inner.fmt(fmt),
+            &Error::EncoderError(ref inner) => inner.fmt(fmt),
+            &Error::DecoderError(ref inner) => inner.fmt(fmt),
+            &Error::IoError(ref inner) => inner.fmt(fmt),
+            &Error::ArgumentError(ref inner) => inner.fmt(fmt),
+            &Error::OperationError(ref inner) => inner.fmt(fmt),
+            &Error::ResponseError(ref inner) => inner.fmt(fmt),
+            &Error::CursorMissingError => write!(fmt, "No cursor found"),
```

Comment:
Let's standardize our error message reporting. Sometimes we end with period and sometimes don't.